### PR TITLE
Controlled casting

### LINKED ISSUE #126: Controlled Casting
**Problem**

In the course of implementing #92 , I've realized that the default casting in PostgreSQL doesn't quite serve our needs.  For example, many different strings cast to booleans, but this loses information.  Another example:  Numeric 1s and 0s _can't_ cast to boolean.  When casting `NUMERIC` to `INTEGER` values are silently rounded.  And so on.

**Proposed solution**

We need to define functions for all supported casting to be used in altering column types and column inference.

**Additional context**

This will probably involve a number of PRs and sub-issues.

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #126 

This adds a number of DB functions, and replaces the column alteration logic to use them.  These functions give us more precise control over what is done when casting one type to another (mostly to make things safer).  

**Technical details**

This requires knocking out your Mathesar DB and reinstalling, since we still don't have upgrade functionality working.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/types/alteration.py
Comment (kgodey) [inline db/types/alteration.py:13]: Datetimes should be fairly simple to add too, right?

File: db/types/alteration.py
Comment (kgodey) [inline db/types/alteration.py:59]: Could you document what this function (and other similar functions) are doing in comments?

**Review by kgodey — APPROVED** [review]
@mathemancer I added some comments but I'm approving to unblock merge. If I'm around when you're done making changes, feel free to ask me to rereview before merge.

File: db/types/alteration.py
Comment (mathemancer) [reply db/types/alteration.py:13]: I'd rather leave those for a separate issue / PR, since there'll be some complexity around dates vs times vs datetimes, and TZs (how to decide whether a column of times should be imported with the local TZ, assuming UTC, etc.)

Comment (mathemancer) [reply db/types/alteration.py:13]: I know we've discussed the plan w.r.t. TZs, but I think having dates/times/datetimes separated out in a well-documented PR might benefit future code archaeologist.

File: db/types/alteration.py
Comment (mathemancer) [reply db/types/alteration.py:59]: Good idea.

Comment (mathemancer) [reply db/types/alteration.py:59]: I made them docstrings to make it visible using the `help` function.  I also considered documenting them in comments in the `PL/pgSQL` functions themselves instead, but I think folks are more likely to be able to (or want to) read the python docstrings than comments in the SQL.  Do you agree?

Comment (kgodey) [reply db/types/alteration.py:59]: Yes, agreed.

File: db/types/alteration.py
Comment (kgodey) [reply db/types/alteration.py:13]: Makes sense.

**Review by kgodey — APPROVED** [review]
Looks great, @mathemancer

**Pull request merged**
